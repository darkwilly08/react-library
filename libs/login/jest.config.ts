module.exports = {
  displayName: 'login',
  preset: '../../jest.preset.ts',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/login'
};
