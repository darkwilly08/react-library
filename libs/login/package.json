{
  "name": "@darkwilly08/login",
  "version": "0.2.5-alpha.0"
}
