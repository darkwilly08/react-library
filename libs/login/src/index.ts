export * from './lib/login';
export * from './lib/hooks/useAuth';
export * from './lib/contexts/userContext';
