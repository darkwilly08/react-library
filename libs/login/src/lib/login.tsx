/* eslint-disable @typescript-eslint/no-non-null-assertion */
import styles from './login.module.scss';
import * as PropTypes from 'prop-types';

import { DwButton } from '@darkwilly08/dw-button';

import TextField from '@mui/material/TextField';
import InputAdornment from '@mui/material/InputAdornment';
import { useState } from 'react';

import { MouseEvent } from 'react';

const propTypes = {
  image: PropTypes.string.isRequired,
  logo: PropTypes.string,
  title: PropTypes.string,
  onLogin: PropTypes.func.isRequired,
  labels: PropTypes.shape({
    username: PropTypes.string,
    password: PropTypes.string,
    login: PropTypes.string,
  }),
};

export type LoginProps = PropTypes.InferProps<typeof propTypes>;

interface Field {
  value: string;
  error: string | null;
}

export function Login({ image, logo, title, onLogin, labels }: LoginProps) {
  const [username, setUsername] = useState<Field>({
    value: '',
    error: null,
  });
  const [password, setPassword] = useState<Field>({
    value: '',
    error: null,
  });
  const [passwordVisibility, setPasswordVisibility] = useState(false);

  const usernameValidator = (value: string) => {
    if (!value) {
      return 'This field is mandatory';
    }

    const usernameLength = 6;
    const patterNoSpaces = '^\\S*$';

    if (value.length < usernameLength) {
      return `Ingrese al menos ${usernameLength} caracteres`;
    } else if (!new RegExp(patterNoSpaces).test(value)) {
      return 'Remueva los espacios en blanco';
    }

    return null;
  };

  const passwordValidator = (value: string) => {
    if (!value) {
      return 'This field is mandatory';
    }

    const passwordLength = 8;
    const patternAtleasUpperCase = '^(.*[A-Z].*)$';
    const patternAtleastDigit = '^(.*[0-9].*)$';

    if (value.length < passwordLength) {
      return `Ingrese al menos ${passwordLength} caracteres`;
    } else if (!new RegExp(patternAtleastDigit).test(value)) {
      return 'Ingrese al menos un dígito';
    } else if (!new RegExp(patternAtleasUpperCase).test(value)) {
      return 'Ingrese al menos una mayúscula';
    }

    return null;
  };

  const handlePasswordVisibility = (e: MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.preventDefault();
    setPasswordVisibility((prevValue) => !prevValue);
  };

  const handlePasswordChange = (value: string) => {
    console.log('onpasswordchangr', value);
    setPassword({
      value,
      error: passwordValidator(value),
    });
  };

  const handleUsernameFocus = () => {
    setUsername((prevState) => {
      return {
        value: prevState.value,
        error: null,
      };
    });
  };
  const handleUsernameBlur = (value: string) => {
    setUsername((prevState) => {
      return {
        value: prevState.value,
        error: usernameValidator(value),
      };
    });
  };
  const handleUsernameChange = (value: string) => {
    setUsername((prevState) => {
      return {
        value,
        error: prevState.error,
      };
    });
  };

  return (
    <div className={styles['container']}>
      <div className={styles['login-card']}>
        <div className={styles['login-card__image']}>
          <img src={image} alt="company" />
          {logo && (
            <div className={styles['login-card__image__logo']}>
              <img src={logo} alt="company" />
            </div>
          )}
        </div>
        <form
          className={styles['login-card__form']}
          onSubmit={(e) => {
            e.preventDefault();
            onLogin(username.value, password.value);
          }}
        >
          <div className="text-h6">{title}</div>
          <div className={styles['login-card__form__fields-container']}>
            <TextField
              value={username.value}
              label={labels?.username}
              onChange={(evt) => handleUsernameChange(evt.target.value)}
              onBlur={(evt) => handleUsernameBlur(evt.target.value)}
              onFocus={handleUsernameFocus}
              variant="filled"
              autoComplete="username"
              error={!!username.error}
              helperText={username.error}
            />
            <TextField
              value={password.value}
              label={labels?.password}
              onChange={(evt) => handlePasswordChange(evt.target.value)}
              type={passwordVisibility ? 'text' : 'password'}
              autoComplete="current-password"
              error={!!password.error}
              helperText={password.error}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <DwButton
                      onClick={handlePasswordVisibility}
                      flat
                      round
                      icon={passwordVisibility ? 'visibility' : 'visibility_off'}
                    />
                  </InputAdornment>
                ),
              }}
              variant="filled"
            />
          </div>
          <DwButton fullWidth text={labels?.login} onClick={() => null} />
        </form>
      </div>
    </div>
  );
}

Login.propTypes = propTypes;

Login.defaultProps = {
  image: 'https://dummyimage.com/700x350/000/fff',
  logo: 'https://dummyimage.com/60x60/d6a3d6/000000',
  title: 'App name',
  labels: {
    username: 'Username',
    password: 'Password',
    login: 'login',
  },
};
