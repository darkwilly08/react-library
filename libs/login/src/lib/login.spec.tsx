import { render } from '@testing-library/react';

import { Login, LoginProps } from './login';

describe('Login', () => {
  let loginProps: LoginProps;
  beforeEach(() => {
    loginProps = {
      title: 'Example App',
      onLogin: jest.fn(),
      image: 'https://dummy.com',
    };
  });
  it('should render successfully', () => {
    const { container } = render(
      <Login title={loginProps.title} onLogin={loginProps.onLogin} image={loginProps.image} />,
    );
    expect(container).toMatchSnapshot();
  });
});
