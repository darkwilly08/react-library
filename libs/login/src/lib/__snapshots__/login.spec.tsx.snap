// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render successfully 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="login-card"
    >
      <div
        class="login-card__image"
      >
        <img
          alt="company"
          src="https://dummy.com"
        />
        <div
          class="login-card__image__logo"
        >
          <img
            alt="company"
            src="https://dummyimage.com/60x60/d6a3d6/000000"
          />
        </div>
      </div>
      <form
        class="login-card__form"
      >
        <div
          class="text-h6"
        >
          Example App
        </div>
        <div
          class="login-card__form__fields-container"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-1"
              id="mui-1-label"
            >
              Username
            </label>
            <div
              class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-19mk8g1-MuiInputBase-root-MuiFilledInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="username"
                class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
                id="mui-1"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-2"
              id="mui-2-label"
            >
              Password
            </label>
            <div
              class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-g9m4e0-MuiInputBase-root-MuiFilledInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiFilledInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1gctnaj-MuiInputBase-input-MuiFilledInput-input"
                id="mui-2"
                type="password"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-jv61v3-MuiInputAdornment-root"
              >
                <div
                  class="react-ripples ripple-button--round"
                  style="position: relative; display: inline-flex; overflow: hidden;"
                >
                  <button
                    class="ripple-button__button ripple-button__button--flat ripple-button__button--round"
                  >
                    <span
                      class="material-icons ripple-button__button__left-icon ripple-button__button__left-icon--round"
                    >
                      visibility_off
                    </span>
                  </button>
                  <s
                    style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="react-ripples ripple-button ripple-button--full-width"
          style="position: relative; display: inline-flex; overflow: hidden;"
        >
          <button
            class="ripple-button__button ripple-button__button--full-width"
          >
            <div>
              login
            </div>
          </button>
          <s
            style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none;"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
